File: kubik/kubik-server/src/main/java/com/cspinformatique/kubik/server/domain/product/service/ProductService.java
package com.cspinformatique.kubik.server.domain.product.service;

import java.util.List;

import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;

import com.cspinformatique.kubik.server.domain.dilicom.model.Reference;
import com.cspinformatique.kubik.server.model.product.Category;
import com.cspinformatique.kubik.server.model.product.Product;
import com.cspinformatique.kubik.server.model.product.ProductDoubles;
import com.cspinformatique.kubik.server.model.product.Supplier;

public interface ProductService {
	Product buildProductFromReference(Reference reference);

	int countByCategory(Category category);
	
	int countByImagesValidated(boolean imagesValidated);
	
	List<Integer> findAllIds();
	
	List<ProductDoubles> findAllProductDoubles();
	
	Product findByEan13(String ean13);
	
	List<Product> findByCategory(Category category);
	
	Product findByEan13AndSupplier(String ean13, Supplier supplier);

	List<Product> findByEan13Doubles(String ean13);
	
	Iterable<Product> findBySupplier(Supplier supplier);
	
	Page<Product> findAll(Pageable pageable);
	
	Product findOne(int id);
	
	Product findRandomByCategory(Category category);
	
	Product findRandomByImagesValidated(boolean imagesValidated);
	
	void mergeProduct(Product sourceProduct, Product targetProduct);
	
	Product save(Product product);
	
	Product save(Product product, boolean skipBroadleafNotification);
	
	Page<Product> search(String query, Pageable pageable);
}

File: kubik/kubik-server/src/main/webapp/resources/js/product/product-service.js
(function(){
	var PRODUCT_URL = "/product";
	
	angular
		.module("Kubik")
		.factory("productService", ProductService);
	
	function ProductService($http){
		return {
			findByEan13 : findByEan13
		};
		
		function findByEan13(ean13){
			return $http
				.get(PRODUCT_URL + "?ean13=" + ean13)
				.then(findByEan13Success);
			
			function findByEan13Success(response){
				return response.data;
			}
		}
	}
})();